### ISSUE TITLE
Preserve ordering in lists

### ISSUE THREAD

**Issue by CthulhuDen** [sol/hpack#324]
For whatever reason, `cabal new-repl` requires `c-files` to be in dependency order. However, currently hpack breaks the ordering (cabal file has all entries in alphabetic order). Am I missing some option to preserve custom ordering for `c-files` field?

**Comment by sol** [comment]
@CthulhuDen Hey, thanks a lot for reporting this.  This sounds critical, so we want to address this fast.  Do you have any code that you can share which reproduces the issue?

Also, can you give me the output of `cabal --version`.

As a quick workaround you can use something like this for now:

```yaml
library:
  verbatim:
    c-sources: |
      foo.c
      bar.c
      baz.c
```

**Comment by CthulhuDen** [comment]
Thanks for responding! I've set up an example repo: https://github.com/CthulhuDen/cbits-ordering.

After executing `hpack`, running `cabal new-repl` (btw I used cabal 2.4.0.0 and ghc 8.4.3) on clean project loads ghci and then when evaluating `greeting` fails with

    ghc: ^^ Could not load 'greeting', dependency unresolved. See top entry above.
    
    
    ByteCodeLink: can't find label
    During interactive linking, GHCi couldn't find the following symbol:
      greeting


`cabal new-build` succeeds, and running `cabal new-repl` in built project can't load and instead results in

    ghc: panic! (the 'impossible' happened)
      (GHC version 8.4.3 for x86_64-unknown-linux):
        Loading temp shared object failed: /tmp/ghc22985_0/libghc_1.so: undefined symbol: msg

I, however, I manually rearrance `c-files` entries in `cbits-ordering.cabal`, it both builds and loads repl just fine.

The workaround you provided for now kinda works, only you need to also account for indentation (and ignore hpack warning about missing file):

      c-sources: |
        cbits/the_dependency.c
              cbits/a_dependent.c


Fails during invocation of `hpack` with

**Comment by quasicomputational** [comment]
FWIW, the root cause here is https://ghc.haskell.org/trac/ghc/ticket/13786

**Comment by sol** [comment]
@quasicomputational thanks a lot for the pointer.

**Comment by sol** [comment]
@CthulhuDen thanks for providing this minimal example!  This is certainly helpful!

> The workaround you provided for now kinda works, only you need to also account for indentation (and ignore hpack warning about missing file):

`hpack` should take care of indentation.  The following works just fine for me:

```yaml
  verbatim:
    c-sources: |
      cbits/the_dependency.c
      cbits/a_dependent.c
```
Can you double check?

Also, what exact warnings regarding files do you get?

**Comment by CthulhuDen** [comment]
@sol yes indeed it works. I just didn't pay attention to that `verbatim` key, thought it was a dummy section or something. My mistake :-)

**Comment by sol** [comment]
This will be addressed in `hpack-0.30.0`.  I'll try my best to push this out soon, but there are a couple of other things that should still go into `0.30.0`.
